### PR TITLE
factor out and generalise definition of pred_conj and friends

### PR DESCRIPTION
**Pull request by lsf37**
The idea is to use the pre-defined boolean algebra class to have `pred_conj` (`_ and _`) not only on one-argument predicates `'s => bool`, but on predicates (functions to bool) with any number of arguments. In particular, this eliminates the need for a second `And` constant and specialised rules for that.

In addition, it turns out that we defined this setup multiple times separately in each monad development instead of once in a library.

The new `pred_conj` ends up as just an abbreviation for `inf_class.inf` from boolean algebras (to be precise, from the semilattice class that is used to construct the boolean algebra class).

Contains:

- lib update
- proof update
- a fix for a problem in AutoCorresDoc ROOT that for some reason hasn't triggered until now

### CODE REVIEW

File: lib/Monad_WP/TraceMonad.thy [hunk 1]
```diff
@@ -941,54 +941,17 @@ where
...
-abbreviation(input)
-  toptop :: "'a \<Rightarrow> 'b \<Rightarrow> bool" ("\<top>\<top>")
-where
-  "\<top>\<top> \<equiv> \<lambda>_ _. True"
-
+text \<open> Abbreviations for trivial postconditions (taking three arguments): \<close>
 abbreviation(input)
   toptoptop :: "'a \<Rightarrow> 'b \<Rightarrow> 'b \<Rightarrow> bool" ("\<top>\<top>\<top>")
 where
   "\<top>\<top>\<top> \<equiv> \<lambda>_ _ _. True"
```
Comment (Xaphiosis) [inline lib/Monad_WP/TraceMonad.thy:948]: silly consistency question: why can we get rid of toptop, but not toptoptop?

Comment (lsf37) [reply lib/Monad_WP/TraceMonad.thy:948]: All monad Hoare logics introduced toptop, but only the TraceMonad uses toptoptop, and I didn't want to introduce something into the wider context that we should instead get rid of. 

I haven't figured out yet how to get rid of toptop (and therefore toptoptop) without too much impact on how wp works, that's why it still exists at all. But removing it should be the next step in this particular quest.

File: lib/Fun_Pred_Syntax.thy [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+lemma pred_top_comp[simp]:
+  "\<top> \<circ> f = \<top>"
+  by (simp add: comp_def)
+
+lemma pred_bot_comp[simp]:
+  "\<bottom> \<circ> f = \<bottom>"
+  by (simp add: comp_def)
+
+(* We would get these for free if we could instantiate pred_top/pred_bot to top/bot directly: *)
+lemmas pred_top_left_neutral[simp] =
```
Comment (Xaphiosis) [inline lib/Fun_Pred_Syntax.thy:120]: trivial: recommend newline above this line

Comment (Xaphiosis) [reply lib/Fun_Pred_Syntax.thy:120]: actually, looking further above, it seems like I'd recommend it for most of the comments that outline sections of lemmas... have you considered `text` or actual sections of some kind?

File: lib/Fun_Pred_Syntax.thy [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+abbreviation pred_conj :: "('a \<Rightarrow> 'b::boolean_algebra) \<Rightarrow> ('a \<Rightarrow> 'b) \<Rightarrow> ('a \<Rightarrow> 'b)" where
+  "pred_conj \<equiv> inf"
+
+abbreviation pred_disj :: "('a \<Rightarrow> 'b::boolean_algebra) \<Rightarrow> ('a \<Rightarrow> 'b) \<Rightarrow> ('a \<Rightarrow> 'b)" where
+  "pred_disj \<equiv> sup"
+
+abbreviation pred_neg :: "('a \<Rightarrow> 'b::boolean_algebra) \<Rightarrow> ('a \<Rightarrow> 'b)" where
+  "pred_neg \<equiv> uminus"
+
+(* Lifted True/False: ideally, we'd map these to top/bot, but top/bat are constants and there are
```
Comment (Xaphiosis) [inline lib/Fun_Pred_Syntax.thy:29]: top/bat -> top/bot

File: lib/Fun_Pred_Syntax.thy (hunk 2 above)
Comment (lsf37) [reply lib/Fun_Pred_Syntax.thy:120]: actual (sub)sections might make sense, yes, I'll have a look at that

File: lib/Fun_Pred_Syntax.thy [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+theory Fun_Pred_Syntax
+imports Main
+begin
+
+(* Functions are already instances of Boolean algebras and provide all the standard laws one
+   would like to have. Default simplifications are automatic. Search for inf/sup/uminus to
+   find further laws and/or unfold via the definitions below.
+
+   The abbreviations here provide special syntax for the function instance of Boolean
+   algebras only, leaving other instances (such as "bool") untouched. *)
```
Comment (Xaphiosis) [inline lib/Fun_Pred_Syntax.thy:18]: I'll be honest, I don't really follow what is happening here and the comment doesn't help me that much either, likely due to `boolean_algebra` being a new thing for me.
Reading the header of this theory, "Syntax for using functions as predicates", it would imply that any functions can be used as predicates?
You're excluding "bool". Is that to avoid being able to say "True and True" thus duplicating the normal logical operators?

**Review by Xaphiosis — APPROVED** [review]
As we discussed, the commit message "lib+proofs+sys-init+tools: proof updates for Fun_Pred" should be Fun_Pred_Syntax

I'd recommend tweaking the comments that delineate sections to not be directly above a single lemma, as it looks like that comment only applies to that lemma.

Other than that, this is a very interesting generalisation!

File: lib/Fun_Pred_Syntax.thy (hunk 4 above)
Comment (lsf37) [reply lib/Fun_Pred_Syntax.thy:18]: > You're excluding "bool". Is that to avoid being able to say "True and True" thus duplicating the normal logical operators?

Yes, precisely, I don't want people to start using these for `True and True` for bool, just because it's theoretically possible.

> Reading the header of this theory, "Syntax for using functions as predicates", it would imply that any functions can be used as predicates?

Any predicate-like function type, yes. I.e. `'s => bool` as before or any `'a => 'b => 'c => 'd => .. => bool` (where and of the type variables can themselves be functions or anything else) or any `'x => 'y` where `'y` is a boolean algebra. As long as it ends in bool or a in a boolean algebra, it'll work.

Basically, `'s => bool` is what we'd usually call predicates (functions from something to bool), and this generalises it to any number of arguments where the arguments can be anything and the result doesn't have to be strictly bool, but is bool-like enough to satisfy the laws of booleans (= boolean_algebra).

Comment (Xaphiosis) [reply lib/Fun_Pred_Syntax.thy:18]: Would it make any sense to update the header to reflect that a bit better, if this theory provides syntax for predicate-like functions with an arbitrary number of arguments? I'm ok either way, but speaking from a perspective of having been confused after reading it.

Comment (lsf37) [reply lib/Fun_Pred_Syntax.thy:18]: Yup, can add a bit more to the header.

Comment (Xaphiosis) [reply lib/Fun_Pred_Syntax.thy]: Thumbs up.
